File: app/src/main/java/com/prey/actions/triggers/TriggerController.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2019 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import com.prey.PreyLogger;
import com.prey.json.UtilJson;
import com.prey.net.PreyWebServices;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TriggerController {

    private List<TriggerDto> listBD = null;
    private List<TriggerDto> listWeb = null;

    private static TriggerController instance;

    public static TriggerController getInstance() {
        if (instance == null) {
            instance = new TriggerController();
        }
        return instance;
    }

    public void run(Context ctx) {
        try {
            Thread.sleep(1000);
            TriggerDataSource dataSource = new TriggerDataSource(ctx);
            listBD = dataSource.getAllTriggers();
            listWeb =null;
            try {listWeb = TriggerParse.getJSONFromUrl(ctx); } catch (Exception e) {}
            updateTriggers(ctx,listWeb,listBD,dataSource);
        } catch (Exception e) {
            PreyLogger.e("error run"+e.getMessage(),e);
        }
    }

    private void updateTriggers(Context ctx, List<TriggerDto> listWeb , List<TriggerDto> listBD, TriggerDataSource dataSource){
        try {
            List<TriggerDto> listDelete=new ArrayList<>();
            List<TriggerDto> listUpdate=new ArrayList<>();
            Map<Integer, TriggerDto> mapBD = convertMap(listBD);
            Map<Integer, TriggerDto> mapWeb = convertMap(listWeb);
            List<Integer> removeList = new ArrayList<Integer>();
            List<Integer> listRemove = new ArrayList<Integer>();
            List<TriggerDto> listAdd = new ArrayList<TriggerDto>();
            List<TriggerDto> listRun= new ArrayList<TriggerDto>();
            List<TriggerDto> listDel = new ArrayList<TriggerDto>();
            List<TriggerDto> listStop = new ArrayList<TriggerDto>();
            for(int i=0;listBD!=null&&i<listBD.size();i++){
                TriggerDto dto=listBD.get(i);
                if(mapWeb!=null&&!mapWeb.containsKey(dto.getId())){
                    removeList.add(dto.getId());
                    listRemove.add(dto.getId());
                    listDel.add(dto);
                    dataSource.deleteTrigger(""+dto.getId());
                }
            }
            if(listDel.size()>0) {
                cancelAlarm(ctx,listDel);
            }
            if (removeList != null && removeList.size() > 0) {
                String infoDelete = "[";
                for (int i = 0; removeList != null && i < removeList.size(); i++) {
                    infoDelete += removeList.get(i);
                    if (i + 1 < removeList.size()) {
                        infoDelete += ",";
                    }
                }
                infoDelete += "]";
                PreyLogger.d("Trigger infoDelete:" + infoDelete);
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "stopped", infoDelete));
            }
            for(int i=0;listWeb!=null&&i<listWeb.size();i++) {
                TriggerDto tri = listWeb.get(i);
                if(mapBD.containsKey(tri.getId())){
                    listRun.add(tri);
                }else{
                    listAdd.add(tri);
                }
            }
            for(int i=0;listRun!=null&&i<listRun.size();i++) {
                TriggerDto trigger = listRun.get(i);
                try {
                    TimeTrigger.updateTrigger(ctx, trigger);
                }catch (TriggerException te) {
                    listStop.add(trigger);
                    PreyLogger.d("TimeTrigger listRun exception id:"+trigger.id+" ,state:"+te.code);
                }
            }
            String infoStop = "[";
            for(int i=0;listStop!=null&&i<listStop.size();i++) {
                TriggerDto trigger = listStop.get(i);
                infoStop += "{\"id\":"+trigger.id+",\"state\":3}";
                if (i + 1 < listStop.size()) {
                    infoStop += ",";
                }
                dataSource.deleteTrigger(""+trigger.id);
            }
            infoStop += "]";
            if(listStop!=null&&listStop.size()>0) {
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "started", infoStop));
            }
            String infoAdd = "[";
            for(int i=0;listAdd!=null&&i<listAdd.size();i++) {
                TriggerDto trigger = listAdd.get(i);
                try {
                    TimeTrigger.updateTrigger(ctx, trigger);
                    infoAdd += "{\"id\":"+trigger.id+",\"state\":1}";
                    dataSource.createTrigger(trigger);
                }catch (TriggerException te){
                    infoAdd += "{\"id\":"+trigger.id+",\"state\":"+te.code+"}";
                }
                if (i + 1 < listAdd.size()) {
                    infoAdd += ",";
                }
            }
            infoAdd += "]";
            if(listAdd!=null&&listAdd.size()>0) {
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "started", infoAdd));
            }
        } catch (Exception e) {
            PreyLogger.e("error run"+e.getMessage(),e);
        }
    }

    private Map<Integer, TriggerDto> convertMap(List<TriggerDto> list) {
        if(list==null){
            return null;
        }
        Map<Integer, TriggerDto> map = new HashMap<Integer, TriggerDto>();
        for (int i = 0; i < list.size(); i++) {
            TriggerDto tri = list.get(i);
            map.put(tri.getId(), tri);
        }
        return map;
    }


    public void cancelAlarm(Context ctx, List<TriggerDto> list) {
        AlarmManager alarmManager = (AlarmManager) ctx.getSystemService(ctx.ALARM_SERVICE);
        for (int i = 0; list != null && i < list.size(); i++) {
            TriggerDto trigger = list.get(i);
            String events = trigger.getEvents();
            if (events.indexOf(TimeTrigger.EXACT_TIME) > 0) {
                Intent intent = new Intent(ctx, TimeTriggerReceiver.class);
                intent.putExtra("trigger_id", "" + trigger.id);
                PendingIntent pendingIntent = PendingIntent.getBroadcast(ctx, (trigger.id * TimeTrigger.ADD_TRIGGER_ID), intent, PendingIntent.FLAG_UPDATE_CURRENT);
                alarmManager.cancel(pendingIntent);
            }
            if (events.indexOf(TimeTrigger.REPEAT_TIME) > 0) {
                List<TriggerEventDto> listEvents = TriggerParse.TriggerEvents(events);
                for (int j = 0; listEvents != null && j < listEvents.size(); j++) {
                    TriggerEventDto event = listEvents.get(j);
                    if ("repeat_time".equals(event.type)) {
                        try {
                            JSONObject json = new JSONObject(event.info);
                            JSONArray array = json.getJSONArray("days_of_week");
                            for (int x = 0; array != null && x < array.length(); x++) {
                                int day = array.getInt(x);
                                Intent intent = new Intent(ctx, TimeTriggerReceiver.class);
                                intent.putExtra("trigger_id", "" + trigger.id);
                                PendingIntent pendingIntent = PendingIntent.getBroadcast(ctx, (trigger.id * TimeTrigger.ADD_TRIGGER_ID + day), intent, PendingIntent.FLAG_UPDATE_CURRENT);
                                alarmManager.cancel(pendingIntent);
                            }
                        } catch (Exception e) {
                            PreyLogger.e("Error cancelAlarm:"+e.getMessage(),e);
                        }
                    }
                }
            }
        }
    }

    public void sendNotify(final Context ctx, final Map<String, String> params) {
        new Thread() {
            public void run() {
                try{
                    PreyWebServices.getInstance().sendNotifyActionResultPreyHttp(ctx, params);
                } catch (Exception e) {
                }
            }
        }.start();
    }
}

File: app/src/main/java/com/prey/actions/triggers/kotlin/TriggerController.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers.kotlin

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import com.prey.json.kotlin.UtilJson
import com.prey.kotlin.PreyLogger
import com.prey.net.kotlin.PreyWebServices
import com.prey.net.kotlin.UtilConnection

import org.json.JSONObject

class TriggerController {
    private var listBD: List<TriggerDto>? = null
    private var listWeb: List<TriggerDto>? = null

    fun run(ctx: Context) {
        try {
            if (UtilConnection.getInstance().isInternetAvailable()) {
                Thread.sleep(1000)
                val dataSource = TriggerDataSource(ctx)
                listBD = dataSource.allTriggers
                listWeb = null
                try {
                    listWeb = TriggerParse.getJSONFromUrl(ctx)
                } catch (e: Exception) {
                    PreyLogger.e("error TriggerController get json:" + e.message, e)
                }
                updateTriggers(ctx, listWeb, listBD, dataSource)
            }
        } catch (e: Exception) {
            PreyLogger.e("error TriggerController run:" + e.message, e)
        }
    }

    private fun updateTriggers(
        ctx: Context,
        listWeb: List<TriggerDto>?,
        listBD: List<TriggerDto>?,
        dataSource: TriggerDataSource
    ) {
        try {
            val listDelete: List<TriggerDto> = ArrayList()
            val listUpdate: List<TriggerDto> = ArrayList()
            val mapBD = convertMap(listBD)
            val mapWeb = convertMap(listWeb)
            val removeList: MutableList<Int> = ArrayList()
            val listRemove: MutableList<Int> = ArrayList()
            val listAdd: MutableList<TriggerDto> = ArrayList()
            val listRun: MutableList<TriggerDto> = ArrayList()
            val listDel: MutableList<TriggerDto> = ArrayList()
            val listStop: MutableList<TriggerDto> = ArrayList()
            run {
                var i = 0
                while (listBD != null && i < listBD.size) {
                    val dto = listBD[i]
                    if (mapWeb != null && !mapWeb.containsKey(dto.getId())) {
                        removeList.add(dto.getId())
                        listRemove.add(dto.getId())
                        listDel.add(dto)
                        dataSource.deleteTrigger("" + dto.getId())
                    }
                    i++
                }
            }
            if (listDel.size > 0) {
                cancelAlarm(ctx, listDel)
            }
            if (removeList != null && removeList.size > 0) {
                var infoDelete = "["
                var i = 0
                while (removeList != null && i < removeList.size) {
                    infoDelete += removeList[i]
                    if (i + 1 < removeList.size) {
                        infoDelete += ","
                    }
                    i++
                }
                infoDelete += "]"
                PreyLogger.d("Trigger infoDelete:$infoDelete")
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "stopped", infoDelete))
            }
            run {
                var i = 0
                while (listWeb != null && i < listWeb.size) {
                    val tri = listWeb[i]
                    if (mapBD!!.containsKey(tri.getId())) {
                        listRun.add(tri)
                    } else {
                        listAdd.add(tri)
                    }
                    i++
                }
            }
            run {
                var i = 0
                while (listRun != null && i < listRun.size) {
                    val trigger = listRun[i]
                    try {
                        TimeTrigger.updateTrigger(ctx, trigger)
                    } catch (te: TriggerException) {
                        listStop.add(trigger)
                        PreyLogger.d("TimeTrigger listRun exception id:" + trigger.getId() + " ,state:" + te.code)
                    }
                    i++
                }
            }
            var infoStop = "["
            run {
                var i = 0
                while (listStop != null && i < listStop.size) {
                    val trigger = listStop[i]
                    infoStop += "{\"id\":" + trigger.getId() + ",\"state\":3}"
                    if (i + 1 < listStop.size) {
                        infoStop += ","
                    }
                    dataSource.deleteTrigger("" + trigger.getId())
                    i++
                }
            }
            infoStop += "]"
            if (listStop != null && listStop.size > 0) {
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "started", infoStop))
            }
            var infoAdd = "["
            var i = 0
            while (listAdd != null && i < listAdd.size) {
                val trigger = listAdd[i]
                try {
                    TimeTrigger.updateTrigger(ctx, trigger)
                    infoAdd += "{\"id\":" + trigger.getId() + ",\"state\":1}"
                    dataSource.createTrigger(trigger)
                } catch (te: TriggerException) {
                    infoAdd += "{\"id\":" + trigger.getId() + ",\"state\":" + te.code + "}"
                }
                if (i + 1 < listAdd.size) {
                    infoAdd += ","
                }
                i++
            }
            infoAdd += "]"
            if (listAdd != null && listAdd.size > 0) {
                sendNotify(ctx, UtilJson.makeMapParam("start", "triggers", "started", infoAdd))
            }
        } catch (e: Exception) {
            PreyLogger.e("error run" + e.message, e)
        }
    }

    private fun convertMap(list: List<TriggerDto>?): Map<Int, TriggerDto>? {
        if (list == null) {
            return null
        }
        val map: MutableMap<Int, TriggerDto> = HashMap()
        for (i in list.indices) {
            val tri = list[i]
            map[tri.getId()] = tri
        }
        return map
    }

    fun cancelAlarm(ctx: Context, list: List<TriggerDto>?) {
        val alarmManager = ctx.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        var i = 0
        while (list != null && i < list.size) {
            val trigger = list[i]
            val events = trigger.getEvents()
            if (events.indexOf(TimeTrigger.EXACT_TIME) > 0) {
                val intent = Intent(ctx, TimeTriggerReceiver::class.java)
                intent.putExtra("trigger_id", "" + trigger.getId())
                val pendingIntent = PendingIntent.getBroadcast(
                    ctx,
                    (trigger.getId() * TimeTrigger.ADD_TRIGGER_ID),
                    intent,
                    PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
                )
                alarmManager.cancel(pendingIntent)
            }
            if (events.indexOf(TimeTrigger.REPEAT_TIME) > 0) {
                val listEvents = TriggerParse.TriggerEvents(events)
                var j = 0
                while (listEvents != null && j < listEvents.size) {
                    val event = listEvents[j]
                    if ("repeat_time" == event.getType()) {
                        try {
                            val json = JSONObject(event.getInfo())
                            val array = json.getJSONArray("days_of_week")
                            var x = 0
                            while (array != null && x < array.length()) {
                                val day = array.getInt(x)
                                val intent = Intent(ctx, TimeTriggerReceiver::class.java)
                                intent.putExtra("trigger_id", "" + trigger.getId())
                                val pendingIntent = PendingIntent.getBroadcast(
                                    ctx,
                                    (trigger.getId() * TimeTrigger.ADD_TRIGGER_ID + day),
                                    intent,
                                    PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
                                )
                                alarmManager.cancel(pendingIntent)
                                x++
                            }
                        } catch (e: Exception) {
                            PreyLogger.e("Error cancelAlarm:" + e.message, e)
                        }
                    }
                    j++
                }
            }
            i++
        }
    }

    fun sendNotify(ctx: Context, params: MutableMap<String, String?>) {
        object : Thread() {
            override fun run() {
                try {
                    PreyWebServices.getInstance().sendNotifyActionResultPreyHttp(ctx, params)
                } catch (e: Exception) {
                }
            }
        }.start()
    }

    companion object {
        private var instance: TriggerController? = null
        fun getInstance(): TriggerController {
            if (instance == null) {
                instance = TriggerController()
            }
            return instance!!
        }
    }
}